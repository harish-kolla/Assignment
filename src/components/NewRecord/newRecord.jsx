import React from "react";

import { staticKeys } from "../../constants";
import deleteIcon from "../../icons/delete.svg";

export const AddNewRecord = ({ inputChangeHandler, newRecord, updateRow }) => {
  return (
    <tr data-testid="add-new-row">
      <td>
        <input
          data-testid="add-new-item"
          type="text"
          value={newRecord["item"] || ""}
          onChange={(e) => inputChangeHandler(e, "item")}
        />
      </td>

      <td>
        <input
          data-testid="add-new-itemDesc"
          type="text"
          value={newRecord["itemDesc"] || ""}
          onChange={(e) => inputChangeHandler(e, "itemDesc")}
        />
      </td>
      <td>
        <input
          data-testid="add-new-price"
          type="text"
          value={newRecord["price"] || ""}
          onChange={(e) => inputChangeHandler(e, "price")}
        />
      </td>
      <td>
        <button
          data-testid="add-new-row-remove-button"
          type="button"
          className="btn btn-link"
          onClick={() => updateRow(false)}
        >
          <img src={deleteIcon} alt={staticKeys.remove} />
        </button>
      </td>
    </tr>
  );
};
