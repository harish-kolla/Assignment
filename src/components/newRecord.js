import React from "react";

export const AddNewRecord = ({ inputChangeHandler, newRecord, updateRow }) => {
  return (
    <tr>
      <td>
        <input
          type="text"
          value={newRecord["item"] || ""}
          onChange={(e) => inputChangeHandler(e, "item")}
        />
      </td>

      <td>
        <input
          type="text"
          value={newRecord["itemDesc"] || ""}
          onChange={(e) => inputChangeHandler(e, "itemDesc")}
        />
      </td>
      <td>
        <input
          type="text"
          value={newRecord["price"] || ""}
          onChange={(e) => inputChangeHandler(e, "price")}
        />
      </td>
      <td>
        <button
          type="button"
          className="btn btn-link"
          onClick={() => updateRow(false)}
        >
          Remove
        </button>
      </td>
    </tr>
  );
};
