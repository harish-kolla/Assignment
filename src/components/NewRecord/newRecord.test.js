import { render, fireEvent } from "@testing-library/react";

import { AddNewRecord } from "./newRecord";

const testProps = {
  newRecord: {
    id: 0,
    item: "item no 0",
    itemDesc: "item description of 0",
    price: 100,
  },
  inputChangeHandler: jest.fn(),
  updateRow: jest.fn(),
};

const hasInputValue = (getByDisplayValue, e, inputValue) => {
  return getByDisplayValue(inputValue) === e;
};

test("Snapshot of AddNewRecord", () => {
  const { container } = render(<AddNewRecord {...testProps} />);
  expect(container.firstChild).toMatchSnapshot();
});

test("Should render components", () => {
  const { queryByTestId } = render(<AddNewRecord {...testProps} />);
  expect(queryByTestId("add-new-item")).toBeTruthy();
  expect(queryByTestId("add-new-itemDesc")).toBeTruthy();
  expect(queryByTestId("add-new-price")).toBeTruthy();
  expect(queryByTestId("add-new-row-remove-button")).toBeTruthy();
});

test("Should have currect input values", () => {
  const { queryByTestId, getByDisplayValue } = render(
    <AddNewRecord {...testProps} />
  );
  expect(
    hasInputValue(
      getByDisplayValue,
      queryByTestId("add-new-item"),
      testProps.newRecord.item
    )
  ).toBe(true);
  expect(
    hasInputValue(
      getByDisplayValue,
      queryByTestId("add-new-itemDesc"),
      testProps.newRecord.itemDesc
    )
  ).toBe(true);
  expect(
    hasInputValue(
      getByDisplayValue,
      queryByTestId("add-new-price"),
      testProps.newRecord.price
    )
  ).toBe(true);
});

test("Should call inputChangeHandler on change item input", () => {
  const { queryByTestId } = render(<AddNewRecord {...testProps} />);
  fireEvent.change(queryByTestId("add-new-item"), { target: { value: "23" } });
  expect(testProps.inputChangeHandler).toHaveBeenCalledTimes(1);
});

test("Should call inputChangeHandler on change itemDesc input", () => {
  const { queryByTestId } = render(<AddNewRecord {...testProps} />);
  fireEvent.change(queryByTestId("add-new-itemDesc"), {
    target: { value: "23" },
  });
  expect(testProps.inputChangeHandler).toHaveBeenCalledTimes(1);
});

test("Should call inputChangeHandler on change price input", () => {
  const { queryByTestId } = render(<AddNewRecord {...testProps} />);
  fireEvent.change(queryByTestId("add-new-price"), { target: { value: "23" } });
  expect(testProps.inputChangeHandler).toHaveBeenCalledTimes(1);
});

test("Should call updateRow on click remove button", () => {
  const { queryByTestId } = render(<AddNewRecord {...testProps} />);
  fireEvent.click(queryByTestId("add-new-row-remove-button"));
  expect(testProps.updateRow).toHaveBeenCalledTimes(1);
});
