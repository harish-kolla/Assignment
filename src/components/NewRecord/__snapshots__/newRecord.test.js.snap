// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of AddNewRecord 1`] = `
<tr
  data-testid="add-new-row"
>
  <td>
    <input
      data-testid="add-new-item"
      type="text"
      value="item no 0"
    />
  </td>
  <td>
    <input
      data-testid="add-new-itemDesc"
      type="text"
      value="item description of 0"
    />
  </td>
  <td>
    <input
      data-testid="add-new-price"
      type="text"
      value="100"
    />
  </td>
  <td>
    <button
      class="btn btn-link"
      data-testid="add-new-row-remove-button"
      type="button"
    >
      Remove
    </button>
  </td>
</tr>
`;
